### ISSUE TITLE
DIRCON Zwift

### ISSUE THREAD

**Issue by cagnulein** [cagnulein/qdomyos-zwift#756]
> Missed that Wahoo dircon setting when I first looked. Got it now thanks. DIRCON shows up in Zwift. I can connect but no data transfers. Data active on QZ. Tried changing port to 36866 (port that Wahoo adapter is listed as using) which again connects but no data. Bike doesn't move in game until I switch to "Tab S2" (QZ over bluetooth).
> 
[debug-Sun_Apr_3_16_02_43_2022.log](https://github.com/cagnulein/qdomyos-zwift/files/8407170/debug-Sun_Apr_3_16_02_43_2022.log)
![20220403_160904](https://user-images.githubusercontent.com/48216/161484793-f7967af8-6866-491d-8870-a88d2f09b4c3.jpg)
![20220403_160837](https://user-images.githubusercontent.com/48216/161484806-d5a0324d-fb6f-48c3-862c-17099e99fda8.jpg)

> Picture shows connection via Dircon while pedalling. QZ screen showed values.

**Comment by cagnulein** [comment]
@p3g4asus I didn't check the log yet, but if you wanna double check it too you're welcome :)

**Comment by cagnulein** [comment]
@p3g4asus i don't see any "New connection from" in the log so I suspect that Zwift doesn't try at all to connect to QZ.
I guess I will need to try myself with wireshark on the Mac M1, what do you think?

**Comment by p3g4asus** [comment]
> @p3g4asus i don't see any "New connection from" in the log so I suspect that Zwift doesn't try at all to connect to QZ. I guess I will need to try myself with wireshark on the Mac M1, what do you think?

It seems like Zwift is not trying to connect to QZ Dircon tcp server. I suggest you to see if advertisement packets are showing in wireshark (filter them with the string "udp.port eq 5353"). Maybe Zwift does not like the way QZ advertise dircon server? I tested that advertising method with the other apps that support dircon and it worked. But maybe Zwift needs a slightly different format or any additional data?

**Comment by p3g4asus** [comment]
Now that I look at the scrreshots, it seems evident that zwift sees the advertisements. Are you sure that the one you attached is the log recorded when zwift shows that GUI in screenshots? From these screenshots it seems that Zwift connects to QZ tcp server e receives at least the packets containing the supported characteristics.

**Comment by cagnulein** [comment]
yes it's the same thing that I would like to ask to the OP. Anyway I guess it will be faster to check this myself at this point :)

**Comment by IProgrammedOnce** [comment]
The picture of the tablet screen showing connection by dircon but no values corresponds to the attached log file. At the same time QZ was showing values as I pedalled. The heart rate was connected thru QZ. I exited before starting the game because this was a test after changing the port. After restarting QZ on the new port there was still no data transfer. Then QZ tablet was restarted in case that was a problem. Images and log are from after the restart.

Earlier I had connected on the default QZ port with the same result of no data. To ride I had to re-pair using bluetooth and then everything worked.

**Comment by cagnulein** [comment]
ok thanks @IProgrammedOnce i will give it a try tonight/tomorrow! Thanks!

**Comment by IProgrammedOnce** [comment]
For this run I woke up the Kickr, Rizer, and HRM. Then I started QZ. Once QZ populated I started Zwift. Zwift found the dircon and selected without me. I pedaled but the data showing in QZ did not show in Zwift device selection screen. I cleared the selections and connected via bluetooth to GalaxyTab S2 which is the tablet running QZ and then the values displayed both in QZ and on the Zwift device selection screen.


[debug-Mon_Apr_4_05_32_28_2022.log](https://github.com/cagnulein/qdomyos-zwift/files/8409802/debug-Mon_Apr_4_05_32_28_2022.log)

**Comment by cagnulein** [comment]
thanks @IProgrammedOnce yes the problem is always the same: zwift doesn't ask to QZ to create a connection, so the send of the data doesn't start. I have to check with wireshark which is the issue

**Comment by cagnulein** [comment]
@p3g4asus @IProgrammedOnce the good news is that i'm seeing the same error too. If you want I collected a wireshark dump about it. I didn't check yet because #441 completely ate my brain this evening ;) 
[zwift dircon.pcapng.zip](https://github.com/cagnulein/qdomyos-zwift/files/8413041/zwift.dircon.pcapng.zip)

**Comment by cagnulein** [comment]
@p3g4asus could you also post, if you have it, a working wireshark file? because i can compare it easily with a zwift one. Thanks!

**Comment by IProgrammedOnce** [comment]
Once my Wahoo dircon adapter arrives I will confirm Zwift is actually working. They know Android 11 and 12 are not. Perhaps Zwift running on  Android 9 doesn't work either (Galaxy Tab 3 that I use to run Zwift is Android 9).

Looking at this the problem is most likely Zwift 
https://forums.zwift.com/t/cant-find-wahoo-direct-connect-android-march-2022/581717

**Comment by cagnulein** [comment]
great @IProgrammedOnce 
today I will dig into this more. Yesterday I was completely burned ot. 
I saw now that the adaptor costs 100€ so a raspi0 with QZ costs just 10$, so I guess it's lot cheaper :)

**Comment by IProgrammedOnce** [comment]
I had already ordered it before I saw your note a few days ago saying not to bother. Pehaps I can use it to connect the Kickr to QZ or at least for testing if Zwift even works with dircon yet.

**Comment by cagnulein** [comment]
@IProgrammedOnce i tried the iOS version of QZ with RGT and Dircon works perfectly. So I guess Zwift did a different dircon implementation. All my hope are in your dircon official device :)

**Comment by cagnulein** [comment]
<img width="572" alt="Screenshot 2022-04-06 at 11 55 48" src="https://user-images.githubusercontent.com/48216/161949507-acb20ef1-3100-472f-9a83-c3fcb07bb988.png">
dircon mac address, thanks https://www.youtube.com/watch?v=XtIM5675dLo&t=342s

**Comment by cagnulein** [comment]
<img width="370" alt="Screenshot 2022-04-06 at 14 13 40" src="https://user-images.githubusercontent.com/48216/161972111-5f8d21c1-d782-4090-8e68-c44c784c5042.png">
thanks to https://www.youtube.com/watch?v=DBoCCrYbDi8

**Comment by IProgrammedOnce** [comment]
I wonder if this is still accurate?    https://support.wahoofitness.com/hc/en-us/articles/4402567268754-Connecting-a-KICKR-V5-using-Direct-Connect-DIRCON-

![Screenshot_20220407-204331](https://user-images.githubusercontent.com/101629184/162359185-0724f53c-4523-4a14-809a-c223566a5610.jpg)

**Comment by cagnulein** [comment]
yes i saw it too in the zwift forum but i don't understand what it means

**Comment by IProgrammedOnce** [comment]
My Wahoo dircon adapter arrived. I saw it once in Zwift on Android. I restarted Zwift and after multiple attempts of restarting tablets and apps could not see it again. It does show up as discoverable in the Wahoo app on my Android phone. For a while it did not and power cycling the Kickr seemed to be required. Un/replugging the dircon did not make it reappear. 

In RGT on android I only see Kickr via bluetooth. 

--------------------------
Iphone on everything below.

Picture 1 is options available for power. Note the different icon between QZ and real dircon adapter.
![20220407_212134](https://user-images.githubusercontent.com/101629184/162364271-6a4938c3-b1a6-459b-ba1b-4b86788862c4.jpg)

Second image is options in controllable with QZ dircon selected in power
![20220407_212242](https://user-images.githubusercontent.com/101629184/162364234-b7ee36c4-00ab-4637-8262-8aa371df24c0.jpg)

Third picture is options for cadence with QZ dircon selected in power
![20220407_212311](https://user-images.githubusercontent.com/101629184/162364209-82113dc6-ceec-4f3b-898c-63fb3fa2ec7d.jpg)


In Zwift on iPhone the only thing that shows up is the dircon. Obviously no Ant but it appears if the dircon is available there is no option to use bluetooth. This is without QZ running.

With QZ going the iphone shows both the real and QZ dircon, and the QZ bluetooth.  Note the different icon shown for QZ and the longer identifier.

Selecting the QZ dircon does not give the dircon option on controllable and no data appears for power or cadence while pedalling.

Selecting QZ bluetooth works.

If I select the real dircon in power the controllable options are real dircon or QZ bluetooth. In cadence real dircon, QZ dircon, and QZ bluetooth.

**Comment by cagnulein** [comment]
wow! ok a lot of work. i already understood the different icon. i need to change the advertised uuid to make zwift change the icon.
@IProgrammedOnce is it possible to collect some wireshark sniff about the WDC?
It will be very useful

**Comment by cagnulein** [comment]
also I see that qz is sending 6 bytes as id instead of 5

**Comment by IProgrammedOnce** [comment]
Note a couple edits I made above to clarify.

Ok, I used Wireshark once ages ago. It can be running on any of the devices?  Off to download.

**Comment by cagnulein** [comment]
@IProgrammedOnce yes for the advertisement part you can wireshark anywhere. you can also use the discover apps as in my above screenshot

**Comment by IProgrammedOnce** [comment]
Wireshark appears to be Windows only? and it takes forever to install.  I probably should have restarted the laptop before installing. WindowsMemoryLeaks, a feature not a flaw!!!!!! 16G ram and probably page filing constantly.

Discover apps?

Ok what do you need me to try to capture? I can usually stumble through but I learned to hate computers decades ago when I was messing with a home server on Linux.

**Comment by cagnulein** [comment]
@IProgrammedOnce wireshark is available on windows, linux or mac. It's a very light app so it should be simple to install. Yeah I know the WindowsMemoryLeaks myself when i'm playing with VMs! 

On Wireshark just caputre all the traffic, then I will filter it myself. Thanks

Discover app https://apps.apple.com/us/app/discovery-dns-sd-browser/id305441017 
It's avaiable also on PC and Android if I remember correctly

**Comment by IProgrammedOnce** [comment]
I should probably turn off the TV with Youtube lol. Could grow the filesize just a bit. 

What do you want me to do while recording? Power up the Kickr, start Zwift, etc?

Meaning to ask also can QZ connect to the Kickr via dircon or is it only to the cycling app? I did not see an option to select.

I think this is what you need? Easiest was just using the iphone.

![20220407_232430](https://user-images.githubusercontent.com/101629184/162376980-eeadf39d-1551-46eb-8c92-11801520d0da.jpg)

**Comment by cagnulein** [comment]
> What do you want me to do while recording? Power up the Kickr, start Zwift, etc?

for now just power up the Kickr, it should be enough

> Meaning to ask also can QZ connect to the Kickr via dircon or is it only to the cycling app?

actually just to the cycling app, but we can also manage to connect to the Kickr

> I think this is what you need? Easiest was just using the iphone.

yes, a wireshark log could be useful too :)

**Comment by IProgrammedOnce** [comment]
I can send a file but I am not seeing 192.168.86.80 on any of the packets.

**Comment by cagnulein** [comment]
try to  filter them with the string "udp.port eq 5353" @IProgrammedOnce

**Comment by IProgrammedOnce** [comment]
I am not seeing any traffic on that IP but there must be something.  The Wahoo app sees it.

Any concerns to be had posting the Wireshark publicly?


![20220408_002057](https://user-images.githubusercontent.com/101629184/162385062-67e58b0d-8c57-4586-a433-fbe4f9eed2e8.jpg)
![20220408_002104](https://user-images.githubusercontent.com/101629184/162384975-66730a97-da6c-46f5-b446-3600facae06a.jpg)

**Comment by cagnulein** [comment]
@IProgrammedOnce if you want you can send me the wireshark log to roberto.viola83 at gmail.com
Thanks

**Comment by IProgrammedOnce** [comment]
Ok eq means == got it now. Nothing on port 5353
The only way I see any traffic on ip.addr==192.168.86.80 is if I ping it or try to telnet. Telnet just gives a blank box but I think it does connect based on this picture of the app. When I closed the telnet session it changed per the second picture.


![20220408_014341](https://user-images.githubusercontent.com/101629184/162400383-be1aba98-eafc-45d5-96e4-79915b9385db.jpg)

![20220408_014932](https://user-images.githubusercontent.com/101629184/162400868-25b406ee-84b9-4e0b-93e9-075e67f3bc49.jpg)

**Comment by cagnulein** [comment]
@IProgrammedOnce maybe the switch where your wireshark device is connected is filtering this port I guess. Anyway I will try to do some changes in the weekend based on your discoveries and let's see how it will go. Thanks!

**Comment by IProgrammedOnce** [comment]
All devices connect by wifi mesh. This includes Wireshark on Windows, the iphone that successfully connects to dircon, and my Android phone that the Wahoo app is on, as well as the Android tablets that usually run QZ and Zwift. The tablets seem unable to find the real dircon in Zwift.

You have a couple files to look at.

**Comment by cagnulein** [comment]
> Would a firewall block it? I did not see anything about firewalls in the setup?

Yes absolutely. Leave it disabled on windows.

I agree with you, on the wireshark logs there is notthing so i guess the firewall is handling them

**Comment by cagnulein** [comment]
@IProgrammedOnce do you have also a MAC? I think zwift is not connecting to QZ because it's waiting FOR a connection FROM QZ. The only way to understand this is with wireshark on a mac and running zwift on a mac. Let me know
Or otherwise we have to wait the Zwift DIRCON release on PC (when will it be available?)

**Comment by IProgrammedOnce** [comment]
I don't have a Mac for MacOS. Linux or Windows plus Android or iOS. 

I installed SYSTM on my laptop but ran out of time to connect. The interface was far more concerned with trying to sell me a subscription than making it easy to use. I am at about day 3 of a 14 day trial. I work the next couple days but will try to figure out the connection via SYSTM. If not my next day off is Tuesday.

**Comment by cagnulein** [comment]
@IProgrammedOnce ok so I guess we have to wait the Zwift release for PC or I have to buy myself a DIRCON adapter if I will find it on Amazon so I can return it. I say so because I guess SYSTM app will work with QZ ( @p3g4asus already tested it if i'm right ) so the issue is just about the implementation on Zwift. Let me check on amazon

**Comment by cagnulein** [comment]
[wireshark_capture_bluetooth_on_just_dircon.pcapng.zip](https://github.com/cagnulein/qdomyos-zwift/files/8476050/wireshark_capture_bluetooth_on_just_dircon.pcapng.zip)
very important sniff file from wireshark. tomorrow i will try to align the QZ dircon to this one! thanks @IProgrammedOnce !

**Comment by doudar** [comment]
Looks like Zwift just rolled out Dircon for windows today.

**Comment by IProgrammedOnce** [comment]
Here is a sniff from a dircon ride on Zwift Windows
[Wireshark_dircon_zwift_windows2.pcapng.zip](https://github.com/cagnulein/qdomyos-zwift/files/8485670/Wireshark_dircon_zwift_windows2.pcapng.zip)

**Comment by IProgrammedOnce** [comment]
Here is a sniff from Kickr connected to QZ on a tablet via bluetooth, QZ connected to Windows via dircon for power and cadence. QZ connected to Windows via bluetooth for HR, Rizer steering, and controllable (no option for dircon shown with QZ dircon selected for power). Values displayed on Windows for steering and HR, but NOT for power or cadence. No resistance control that I can test.

QZ on 27, Windows on 29, actual dircon adapter on 80  Wireshark search string for accompanying file 
ip.addr==192.168.86.80 || ip.addr==192.168.86.27 || udp.port==5353 
Terminated before a ride started because no power=no movement

[Wireshark_Kickr_QZdircon_ZwiftWindows.pcapng.zip](https://github.com/cagnulein/qdomyos-zwift/files/8486098/Wireshark_Kickr_QZdircon_ZwiftWindows.pcapng.zip)

**Comment by cagnulein** [comment]
great news @IProgrammedOnce ! yesterday i got issue with my dog so i didn't do anything. i will try to reserve tile for today! thank you so much for the tests!

**Comment by IProgrammedOnce** [comment]
Connecting to the new version of Zwift for Android 1.24.0 QZdircon does not pass values for power or cadence. (expected since this is the same result as Zwift Windows). Unlike Windows if I choose QZdircon for power the option for QZdircon appears for controllable. 
The Wahoo dircon adapter now works with Zwift Android. Of course this alone is not that useful as Kickr must connect to QZ (via bluetooth so far) in order for the Rizer to function properly. Alas Wahoo cannot follow standards. Go QZ! I thank you so much for making my Rizer function properly.

I look forward to when Kickr can connect to QZ via wifi (dircon) but these changes must wait as your family (including the 4 legged furry members) is most important. I hope all are well soon.

**Comment by cagnulein** [comment]
@IProgrammedOnce i will be on vacation the next week and i will have a lot of spare time with my mac, so i guess this will be my first toy :)

**Comment by cagnulein** [comment]
@doudar do you want to implement DIRCON also in SS2K?

**Comment by cagnulein** [comment]
ok i'm collecting here the differences.
QZ is actually has a bug in the advertising

![image](https://user-images.githubusercontent.com/48216/163340213-50dc9ca8-60af-44c4-bc6c-451ea0bbd577.png)

it says only 19 bytes but they are more infact the real DIRCON says 48

![image](https://user-images.githubusercontent.com/48216/163340288-76ffc960-9a32-464f-8409-100cc4702355.png)

**Comment by cagnulein** [comment]
QZ is not sending the A packet at all. We need to add this
![image](https://user-images.githubusercontent.com/48216/163340464-8aa46973-c704-414b-a950-fcb0248b2055.png)

**Comment by cagnulein** [comment]
in the SRV frame there is an unwanted "H" in QZ, we can safely remove this
![image](https://user-images.githubusercontent.com/48216/163340808-fb92b534-680c-4aa3-a5f5-5c7b08147c01.png)

**Comment by cagnulein** [comment]
a lot of differences in the TXT too

![image](https://user-images.githubusercontent.com/48216/163341538-c76f75cb-be15-481b-9747-d263a09ae563.png)

- serial number
- mac-address with ":" instead of "-"
- the uuids are in the 16bit HEX code

**Comment by cagnulein** [comment]
IT WORKS @IProgrammedOnce
The issue was that
https://github.com/cagnulein/qdomyos-zwift/issues/756#issuecomment-1098818031

@p3g4asus i did some changes in the mdsn repository, i will do a PR to your branch in the next hours! It's so exciting!

**Comment by cagnulein** [comment]
my branch here https://github.com/cagnulein/qmdnsengine/tree/zwift

**Comment by cagnulein** [comment]
@doudar maybe adding also a ethernet port :D

**Comment by cagnulein** [comment]
@doudar ah you mean an open source implementation of DIRCON with steroids. I like the idea! Consider this: the wahoo code is already available on QZ, so the implementation should be very simple. I think QZ can already do this on a raspberry pi0 for example

**Comment by doudar** [comment]
Yes, a pi0 is a good (cheap) platform for it as well. 

Do you have the wired communication (cable from the wahoo to the pi) figured out as well?

**Comment by cagnulein** [comment]
@doudar nope, only the bluetooth one. But i guess they just use the same protocol over wire

**Comment by IProgrammedOnce** [comment]
![20220414_110657](https://user-images.githubusercontent.com/101629184/163450994-c81102c6-8681-4989-9107-674e88a4be8f.jpg)
![20220414_101853](https://user-images.githubusercontent.com/101629184/163451059-393f9aac-8b66-491e-8257-4bf9cbfd9914.jpg)

**Comment by cagnulein** [comment]
amazing! it's just an serial port isn't it? do you have an ftdi cable? we can sniffer the protocol or maybe someone on internet already did

**Comment by doudar** [comment]
@IProgrammedOnce ... WOW!

**Comment by IProgrammedOnce** [comment]
I am at the limit of my knowledge. I might have something that can act as a RS232 to usb bridge but I don't know it if I do. I just ordered a couple but they are a week off. I repair things like phones but am weak when it comes to software or device interfacing. Time to learn more I guess. I am an avid collector of bits off Ebay that I really do intend to use one day so something that might work could be here without me knowing.

I say it is RS232 because the ST3232B is described as such in the data sheet. 

If I am able to rig something up I am presuming I will need ground and the tx and rx lines. I presume I need to be on the RS232 bus sniffing usb with wireshark on my pc  to read the data both directions silently so as not to interfere with the operation? I have a RS485 adapter but I don't know that it is compatible. The pinouts seem different as there are 5 pins, D+, D-, ground, R+, R- although looking online RS485 should work bidirectional with 2 data lines. Not sure if the voltage levels would match RS232? 

Yes the 12v is always present. The Kickr can be indicator lights off and wake up from starting Zwift. The dircon is always warm so there should be ample current for a pi. Obviously not a very efficient circuit in the dircon.

I guess I should try to order a piO as all I have are Arduino unless they are compatible with QZ?

**Comment by cagnulein** [comment]
> If I am able to rig something up I am presuming I will need ground and the tx and rx lines. I presume I need to be on the RS232 bus sniffing usb with wireshark on my pc to read the data both directions silently so as not to interfere with the operation? I have a RS485 adapter but I don't know that it is compatible. The pinouts seem different as there are 5 pins, D+, D-, ground, R+, R- although looking online RS485 should work bidirectional with 2 data lines. Not sure if the voltage levels would match RS232?

I'm an expert of RS232 communications :) So don't use the RS485, it's a different voltage level interface. You need a RS232 adapator and yes, you can sniff them without interacting with them. You can use a simple putty terminal also for this job.

Also I have another good news: DIRCON works and also contrallable! I'm building the version right now!

**Comment by IProgrammedOnce** [comment]
I tested the beta build on iOS and the updates to the QZ DIRCON implementation worked great. My set-up was Kickr, Rizer, and HRM connected to QZ on the iphone via bluetooth, then DIRCON over wi-fi from QZ to the Android tablet running Zwift. The HRM and Rizer steering both connected first to QZ on the iphone via bluetooth and then to the Android tablet also via bluetooth.

My only problem was at first the iphone was connected to the built in wi-fi on my modem rather than my Google mesh that the rest of the devices were connected to which caused the DIRCON from QZ not to appear on Zwift on the tablet. Just a little gotcha that caused me more hair loss.

Once the usb-to-rs232 adapter arrives I will see if I can figure out how to log the data. Might have a few questions on how to do that.  

Is the longer term plan to have a device that runs QZ and handles the rs232 interface to the Kickr, bluetooth interface for Rizer+HRM, and dircon over wi-fi to Zwift? 

I still like the idea of a rs232 to wifi adapter and dircon client on QZ because it gives the portable visual interface for QZ athough not needing an extra phone or tablet has merits too.

**Comment by IProgrammedOnce** [comment]
I just realized I already have a rs232-to-TTL adapter in the Wahoo adapter. Pins 9 and 10. Not a FTDI brand chip but it should be doing the same thing. Now to figure out how to sniff it.
![Screenshot_20220418-040747_Acrobat for Samsung](https://user-images.githubusercontent.com/101629184/163800347-c304d7dc-5d61-4eaf-9242-c43f53e6b169.jpg)

**Comment by cagnulein** [comment]
> I tested the beta build you provided on iOS and the updates to the QZ DIRCON implementation worked great. My set-up was Kickr, Rizer, and HRM connected to QZ on iphone via bluetooth, then DIRCON over wi-fi from QZ to the Android tablet running Zwift. The HRM and Rizer steering both connected first to QZ on the iphone via bluetooth and then to the Android tablet also via bluetooth.


@IProgrammedOnce great!

> Once the usb-to-rs232 adapter arrives I will see if I can figure out how to log the data. Might have a few questions on how to do that.  

Absolutely! I'm here!

> Is the longer term plan to have a device such as a pi zero W that runs QZ and handles the rs232 interface to the Kickr, bluetooth interface(s) for Rizer+HRM, and dircon over wi-fi to Zwift? 

Absolutely! I will love to to this!

> Edit:  this is my particular set-up but I realize the idea of QZ is to allow a wide variety of devices to interface with multiple softwares, so basically adding an solution for a hardwired interface. How about a usb via OTG interface?

Why not, but let's start understanding the kickr protocol!

> I still like the idea of a rs232 to wifi adapter and dircon client on QZ because it gives the portable visual interface for QZ, athough not needing an extra phone or tablet has merits too. Perhaps I am just liking this because I already have a slightly scratched Wahoo DIRCON adapter already although there is really no way to avoid using bluetooth for Rizer and HRM.

Love it!

**Comment by cagnulein** [comment]
> I just realized I already have a rs232-to-TTL adapter in the Wahoo adapter. Pins 9 and 10. Not a FTDI brand chip but it should be doing the same thing. Now to figure out how to sniff it.
> 
If you have a ttl usb cable you can connect to them without issue!

**Comment by IProgrammedOnce** [comment]
I don't have the usb/ttl cable yet, but should I be able to get the below to work? I get data on Zwift through the dircon link but can't see much data on the pc yet although there is something. I was trying a few different rates until I ran out of time and had to go to work. Does this set up look like it can get the hex/ascii/whatever you need?

![20220418_163931](https://user-images.githubusercontent.com/101629184/163899160-fecb0ed4-864a-43c8-8803-771fd729d960.jpg)
![20220418_163918](https://user-images.githubusercontent.com/101629184/163899275-f9bcafd1-0743-45a2-90fb-b148691f650b.jpg)

**Comment by cagnulein** [comment]
@IProgrammedOnce Are you using the logic analyzer connected to the Rx only isn't it? If so yes I guess you have to see the bits and from them you can also calculate the correct rate.

**Comment by IProgrammedOnce** [comment]
The bottom picture shows the 2 channels, one connected to TX and one to RX. There seems to be a slight difference left to right as if it could be data one way then response. I will test more after work.

What format do you think I am I trying to get to? I believe that program has the ability to convert. The website says the trial version allows full function.

https://support.saleae.com/user-guide/using-logic/using-protocol-analyzers

**Comment by IProgrammedOnce** [comment]
I forgot to ask if there is an Android version of the beta with DIRCON? I like the bigger screen on the tablet.

**Comment by cagnulein** [comment]
> The bottom picture shows the 2 channels, one connected to TX and one to RX. There seems to be a slight difference left to right as if it could be data one way then response. I will test more after work.

zoom on a frame (the little spikes) so we can see the timespan (rate).

> What format do you think I will need to get to provide useful data to you? I believe that program has the ability to convert. The website says the trial version allows full function.

hex code will be perfect

> I forgot to ask if there is an Android version of the beta with DIRCON? I like the bigger screen on the tablet.

i will build it in the next day. I would like also to add #774 in order to have a release candidate for both platform

**Comment by IProgrammedOnce** [comment]
The attached file was created, and viewable with https://www.saleae.com/downloads/

Let me know if that is helping. I think 14400 gives the least framing errors but I am guessing.  You can alter it by editing the setting for each analyzer.

[DirconSerial.sal.zip](https://github.com/cagnulein/qdomyos-zwift/files/8513531/DirconSerial.sal.zip)

**Comment by cagnulein** [comment]
So strange that one channel has so many errors. In the tx to the kickr I see a start char and it's a good sign (0x80) but I guess the baud rate is still not right...maybe some strange baud rate? Not a standard one I mean

**Comment by cagnulein** [comment]
I will play with the software on Thursday

**Comment by IProgrammedOnce** [comment]
Let me know if you would like me to capture with different software. I found I already had the logic analyzer ($12 on Ebay) in a box and the software was the first one I came across that didn't trigger my antivirus. Saleae looks like very nice hardware and software, priced accordingly. I still have a usb-to-serial cable coming as well though I think this is probably even better.

If you touch the 3 dots to the right of each analyzer 'tab' you can edit settings including baud easily. Each channel can analyze at a different rate.

**Comment by IProgrammedOnce** [comment]
I believe this is the chip that interfaces with the Kickr.
Have a look at https://infocenter.nordicsemi.com/pdf/nRF52840_OPS_v0.5.pdf
Start at page 391. I think the baud rate should be a standard rate above 1200 but there might be something with how stop bits are implemented?

**Comment by IProgrammedOnce** [comment]
Channel 0 RXD ( or TX from the Kickr) seems to follow 7 bytes (correct term?), pause about 0.3 seconds, then 0xA2, pause about 0.6 seconds, another 7 bytes, pause, 0xA2, etc.

Channel 1 TXD ( to Kickr ) seems to follow this pattern: 2 bytes, pause about 0.3 seconds, 3 bytes ( the third is 0xFF), pause about 0.6 seconds, repeat.

**Comment by cagnulein** [comment]
i guess there is an issue on the controllable side, i can't see any notification even if the "controllable icon" is selected. @IProgrammedOnce does it work for you? An user sent me this and there is nothing inside https://github.com/cagnulein/qdomyos-zwift/issues/760#issuecomment-1104759912

**Comment by cagnulein** [comment]
> Channel 0 RXD ( or TX from the Kickr) seems to follow 7 bytes (correct term?), pause about 0.3 seconds, then 0xA2, pause about 0.6 seconds, another 7 bytes, pause, 0xA2, etc. Edit: though occasionally a 0x82 replaces a 0xA2
> 
> Channel 1 TXD ( to Kickr ) seems to follow this pattern: 2 bytes, pause about 0.3 seconds, 3 bytes ( the third is 0xFF), pause about 0.6 seconds, repeat.
> 

WOW! perfect @IProgrammedOnce so they are following a standard! So with a ttl cable we can manage them from the raspberry easily!

**Comment by IProgrammedOnce** [comment]
Continuing with the DIRCON topic:

Using ios 2.10.70

Once I start Zwift the ext incline value never changes if using dircon. To explain further, if I am using bluetooth and ext incline is 8.0, but then I change to dircon, ext incline stays as 8.0 regardless of the value changing in Zwift. If Zwift is -5.0 and I change back to bluetooth then the 8.0 will change to -5.0 in the ext incline tile.

If I use bluetooth then as soon as I connect Zwift the ext incline tile changes to zero and during the ride the value matches what is on Zwift and Rizer works. Controllable works.

If I use the Wahoo DIRCON adapter then controllable DOES work. 

The problem seems to be this QZ version does not get the data FROM Zwift if using dircon.

**Comment by cagnulein** [comment]
@IProgrammedOnce fixed, it will be in the 2.10.71 same issue of #659

**Comment by cagnulein** [comment]
@IProgrammedOnce unfortunately my internet connection is broken
<img width="904" alt="Screenshot 2022-04-21 at 22 08 57" src="https://user-images.githubusercontent.com/48216/164544206-a83d2379-fa68-4a1d-a862-3e040c48044c.png">
i will try to upload this tomorrow :(

**Comment by cagnulein** [comment]
@IProgrammedOnce android version 2.10.71 is online ;) (I didn't try myself!)

**Comment by IProgrammedOnce** [comment]
Ok, back to the software side of things.

I am not having success. On QZ 2.10.71 dircon will not control the Rizer or Kickr. 

Android and iOS versions are the same.

Using dircon for data TO Zwift works fine, even HRM.

If I go back into pairing and: pair power to QZxx, then pair controllable to QZxx, then pair power back to QZ dircon then the Rizer and controllable work.  (I seem to have to do this sequence or QZxx does not appear in controllable)  If I go back to pairing and change controllable to QZ dircon then Rizer and controllable stop working. I can repeat this as many times as I want back and forth.

QZxx is bluetooth. While using bluetooth for controllable the values for inclination and ext incline do change. While using QZ dircon for controllable the inclination, and ext incline do not change. 

To clarify, everything in Zwift can be paired via  dircon except controllable. As long as controllable is paired using bluetooth then Rizer and controllable work.

[debug-Fri_Apr_22_23_42_15_2022.zip](https://github.com/cagnulein/qdomyos-zwift/files/8546538/debug-Fri_Apr_22_23_42_15_2022.zip)

**Comment by cagnulein** [comment]
@IProgrammedOnce Thanks, I had a quick look at it and the issue that I found 2 days ago is fixed now: I see the dircon ftms packets correctly parsed now.

So now there is another issue. I will have a look at it after my afternoon ride :)

Vzfit is a drug to me in these days :)

**Comment by cagnulein** [comment]
@IProgrammedOnce did you try to use QZ Dircon disabling the virtual device bluetooth (leaving only the virtual device dircon)? Because it seems that Zwift receives data from DIRCON but it writes to bluetooth. I guess this is the culprit
Let me know

**Comment by IProgrammedOnce** [comment]
@cagnulein 
I tried turning off 'Virtual device bluetooth' (and immediately restarting QZ before starting Zwift). The controllable still does not work and now QZxx does not appear in Zwift so I cannot select bluetooth for QZ power, cadence, or controllable? I feel like I am missing something because you suggested this could work.

The option to connect directly to Kickr via bluetooth is present. Also steering is working which is via bluetooth so I conclude there is not a problem with the Galaxy Tab S3 running Zwift. Options for all the devices appear for Ant+ as well. Steering values appear in QZ so it would appear Rizer steering is connected to QZ and Zwift steering is connecting to QZ.

The other choice 'Wahoo direct connect' is on.

I have the actual Wahoo dircon unplugged.

If I understand this correctly you are able to use dircon to link QZ *regardless of your fitness equipment* because once QZ has the data from the equipment it can link to Zwift over dircon? If this is the case are you able to successfully use dircon bidirectional with the equipment you use? Does this imply the extra QZ connection to the Rizer is affecting something?

![20220423_145729](https://user-images.githubusercontent.com/101629184/164947280-0f01782c-1597-4c62-adb7-4d3637b9127b.jpg)


![20220423_152007](https://user-images.githubusercontent.com/101629184/164947943-3f384470-a3f0-4ca1-bb07-2f0fd842b98e.jpg)

The Wahoo KICKR 0000-2 33 appeared after I restarted QZ mid ride to see if that would change anything. I guess Zwift remembers all the devices until next restart. I often get multiple QZxx bluetooth versions when I restart QZ mid ride but I think this is the first duplicate dircon.

**Comment by IProgrammedOnce** [comment]
I have no idea if this means anything. I don't normally leave the pairing screen open for a couple hours. When I looked there were 7 copies. Can't see it on screen but I guess that is Wahoo KICKR 0000-2-2-2-2-2-2 33

![20220423_185316](https://user-images.githubusercontent.com/101629184/164952613-24204a1c-8a08-498a-a7f5-4381b0d51cfb.jpg)

**Comment by cagnulein** [comment]
> I tried turning off 'Virtual device bluetooth' (and immediately restarting QZ before starting Zwift). The controllable still does not work and now QZxx does not appear in Zwift so I cannot select bluetooth for QZ power, cadence, or controllable? I feel like I am missing something because you suggested this could work.

hah so you were selecting QZxx before when you were trying to use DIRCON? When you want to use DIRCON you have to select the same DIRCON device on all the tiles in ZWIFT in the pairing screen.

> If I understand this correctly you are able to use dircon to link QZ _regardless of your fitness equipment_ because once QZ has the data from the equipment it can link to Zwift over dircon? 

Yes exactly!

> If this is the case are you able to successfully use dircon bidirectional with the equipment you use? 

Yes my setup is mac M1 for QZ and Zwift on iOS. I tested also M1 for QZ and M1 for Zwift

> Does this imply the extra QZ connection to the Rizer is affecting something?

No i guess it's something with the android version of zwift. I have to try this too

> I have no idea if this means anything. I don't normally leave the pairing screen open for a couple hours. When I looked there were 7 copies. Can't see it on screen but I guess that is Wahoo KICKR 0000-2-2-2-2-2-2 33

It sounds to me like a Zwift DIRCON's implementation bug.
Let me try with a similar setup and I will let you know.

**Comment by cagnulein** [comment]
as a reference, this is what I get running QZ on Mac M1 and Zwift on iOS. 
![IMG_F3097EEFC8EC-1](https://user-images.githubusercontent.com/48216/164954790-5791e5d6-0b86-4443-99e9-ebf5358d5bf3.jpeg)

I'm now downloading the zwift version on android

**Comment by IProgrammedOnce** [comment]
@cagnulein 
To clarify, on my S2 tablet I turned off 'Virtual Device Bluetooth', then restarted QZ, then started Zwift, selected Dircon 'Wahoo KICKR 0000 33' for power, controllable, and cadence. For Heartrate I used 'Wahoo HRM-2183' which is QZ dircon. I started riding. Controllable and Rizer did not work. I went back to the pairing screen planning to use QZxx for the controllable but the option was not available even for power. I restarted QZ. Zwift on my S3 tablet disappeared from from the screen but was still running and I brought it back up from background apps. The option for QZxx was still not available but there was the first 'Wahoo KICKR 0000-2 33'. I left to do some other things with that pairing screen open. When I came back was 7 copies and later was 10 copies.

Time to ride now. Turned 'Virtual Device Bluetooth' back on. Restarted QZ. QZxxx was now available. None of the many dircon versions worked so I restarted Zwift. The choices as pictured below work. For heartrate 'Wahoo KICKR 0000 33' also works.

Paused mid-ride with everything working.

![20220423_203300](https://user-images.githubusercontent.com/101629184/164955413-f0fb4439-a091-415b-93b5-4afd0a5c476e.jpg)

**Comment by cagnulein** [comment]
@IProgrammedOnce i just tried the android version of zwift and, as you saw, there is a major issue: Zwift sees the DIRCON virtual device from QZ but it doesn't interact with it. It sounds to me like a zwift issue on android. Are you able to run Zwift also on a PC? Do you have the same experience?

**Comment by cagnulein** [comment]
hah ok so you're using a Android to Android scenario. I didn't test this yet. @p3g4asus told me that on Android it could be a issue on getting the IP from the QT api, so maybe this is the issue. I have to try

**Comment by IProgrammedOnce** [comment]
@cagnulein  On PC there is no QZ? You mean connecting QZ on Android to Zwift on PC? Unfortunately I updated PC Zwift before trying this and the new version won't see ANY dircon not even the real adapter so I am not able to test. 

I will install Zwift on iOS and test that. My internet is 15Mb/s (small town) so I will post again once I have results.

**Comment by cagnulein** [comment]
> On PC there is no QZ? You mean connecting QZ on Android to Zwift on PC?

yes, sorry for the confusion. I mean QZ on android and Zwift on PC. And/or also QZ on iOS and Zwift on PC/Android. I would like to find the issue in your setup. It could be QZ on android or/and Zwift on Android.

> Unfortunately I updated PC Zwift before trying this and the new version won't see ANY dircon not even the real adapter so I am not able to test. UNLESS there is a problem with my setup, but all I know I changed was updating Zwift.

Hah right! I forgot about this

> I will install Zwift on iOS and test QZandroid to Zwift iOS unless you already have?

I can give it a go myself while you will sleeping :)

> My internet is 15Mb/s (small town) so I will post again once I have results.

Same here! I forgot: where are you located?

**Comment by IProgrammedOnce** [comment]
@cagnulein Actually no. I typed that wrong above wow. Getting myself mixed up.

When the Android version came out 2.10.71 I installed it and found the problem. I switched to the iOS version to try it and this discussion is based on the iOS version of QZ.

**Comment by cagnulein** [comment]
great news @IProgrammedOnce I fixed the auto resistance on zwift via dircon! Zwift devs did a terrible job on the dircon implementation @p3g4asus

**Comment by IProgrammedOnce** [comment]
_Zwift devs did a terrible job on the dircon implementation_
< sarcasm>**SHOCKED** everything else is **perfect**< /sarcasm>

**Comment by IProgrammedOnce** [comment]
Yes Rizer and controllable resistance, connected via QZdircon to controllable in Zwift, seem to be working in QZ version 2.10.75 that I tested on ios today.

**Comment by cagnulein** [comment]
perfect thanks @IProgrammedOnce !

**Comment by doudar** [comment]
To file in the FWIW column, Wahoo's own RGT is perfectly happy with not having the "type a" mdns answer at the top of the advertisement. 

Actually, I'm going to file this in the "Zwift has shitty programming" column.

**Comment by cagnulein** [comment]
I guess we can close this. The only issue with zwift and dircon are related to zwift issue. More details here https://forums.zwift.com/t/zwift-does-not-show-wahoo-direct-connect-pairing-option-on-ios-16-october-2022/592239/33
